File: go/2415. Reverse Odd Levels of Binary Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func reverseOddLevels(root *TreeNode) *TreeNode {
    helper(root.Left, root.Right, 0)
    return root
}

func helper(left, right *TreeNode, depth int) {
    if left == nil || right == nil {
        return
    }
    if depth % 2 == 0 {
        left.Val, right.Val = right.Val, left.Val
    }
    helper(left.Left, right.Right, depth+1)
    helper(left.Right, right.Left, depth+1)
}

File: java/2415. Reverse Odd Levels of Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode reverseOddLevels(TreeNode root) {
        helper(root.left, root.right, 0);
        return root;
    }
    private void helper(TreeNode left, TreeNode right, int depth) {
        if (left == null || right == null) return;
        if (depth % 2 == 0) {
            int temp = left.val;
            left.val = right.val;
            right.val = temp;
        }
        helper(left.left, right.right, depth+1);
        helper(left.right, right.left, depth+1);
    }
}
